mkosi/gentoo.py

```diff
@@ -0,0 +1,482 @@
+
+                # sys-apps/systemd -pkcs11
+                # sys-apps/systemd importd lzma
+                sys-apps/systemd homed cryptsetup -pkcs11
+                # MKOSI: usronly
+                sys-apps/systemd repart
+                # sys-apps/systemd -cgroup-hybrid
+                # sys-apps/systemd vanilla
+                # sys-apps/systemd policykit
+                # MKOSI: make sure we're init (no openrc)
+                sys-apps/systemd sysv-utils
+                # sys-fs/lvm2 device-mapper-only -thin
+                """
+            )
+        )
+
+    def mkosi_conf(self, root: Path) -> None:
+        package_env = self.portage_cfg_dir / "package.env"
+        ebuild_sh_env_dir = root / EBUILD_SH_ENV_DIR
+        package_env.mkdir(exist_ok=True)
+        ebuild_sh_env_dir.mkdir(exist_ok=True)
+
+        # apply whatever we put in mkosi_conf to runs invokation of emerge
+        package_env.joinpath("mkosi.conf").write_text("*/*    mkosi.conf\n")
+
+        # we use this so we don't need to touch upstream files
+        # we also use this for documenting build environment as much as possible
+
+        emerge_vars_str = ""
+        for k, v in self.emerge_vars.items():
+            emerge_vars_str += f'{k}="$' + "{" + k + "}" + f' {v}"\n'
+
+        ebuild_sh_env_dir.joinpath("mkosi.conf").write_text(
+            dedent(
+                f"""\
+                    # MKOSI: these were used during image creation...
+                    # and some more! see under package.*/
+                    #
+                    # usrmerge (see all under profile/)
+                    {emerge_vars_str}
+                    """
+            )
+        )
+
+    def invoke_emerge(
+        self,
+        args: CommandLineArguments,
+        root: Path,
+        inside_stage3: bool = True,
+        pkgs: List[str] = list(),
+        actions: List[str] = list(),
+        opts: List[str] = list(),
+    ) -> None:
+        if not inside_stage3:
+            from _emerge.main import emerge_main  # type: ignore
+
+            PREFIX_OPTS: List[str] = list()
+            if "--sync" not in actions:
+                PREFIX_OPTS = [
+                    f"--config-root={root.resolve()}",
+                    f"--root={root.resolve()}",
+                    f"--sysroot={root.resolve()}",
+                ]
+
+            emerge_main(pkgs + opts + PREFIX_OPTS + self.emerge_default_opts + actions)
+        else:
+            cmdline = ["/usr/bin/emerge"]
+
+            cmdline.extend(pkgs)
+            cmdline.extend(self.emerge_default_opts)
+            cmdline.extend(opts)
+            cmdline.extend(actions)
+
+            MkosiPrinter.print_step("Invoking emerg(1) inside stage3")
+            run_workspace_command(
+                args,
+                root,
+                cmdline,
+                network=True,
+                env=self.emerge_vars,
+                nspawn_params=self.DEFAULT_NSPAWN_PARAMS + ["--bind=/sys"],
```

Comment:
Why `/sys`?

Reply:
`bash` does not use `/sys`…

Reply:
yeah, it 's not `bash(1)` it' s `install(1)` and it's not just for `baselayout`, `install(1)` is used everywhere.

```
>>> Failed to install sys-apps/baselayout-2.7-r3, Log file:
>>>  '/var/tmp/portage/sys-apps/baselayout-2.7-r3/temp/build.log'
>>> Jobs: 0 of 1 complete, 1 failed                 Load avg: 1.88, 0.62, 0.22
>>> Extracting info
 * Package:    sys-apps/baselayout-2.7-r3
 * Repository: gentoo
 * USE:        abi_x86_64 amd64 build kernel_linux userland_GNU elibc_glibc
 * FEATURES:   preserve-libs
>>> Extracting sys-apps/baselayout-2.7-r3
 * checking 25 files for package collisions
>>> Merging sys-apps/baselayout-2.7-r3 to /
make -j8 -C /var/tmp/portage/sys-apps/baselayout-2.7-r3/image/usr/share/baselayout DESTDIR= layout-usrmerge
make: Entering directory '/var/tmp/portage/sys-apps/baselayout-2.7-r3/image/usr/share/baselayout'
# Create base filesytem layout
for x in /dev /run /sys /usr/src /boot /etc/profile.d /home /media /mnt /proc /opt /root /usr/local/bin /usr/local/sbin /var/cache /var/empty /var/lib /var/log /var/spool ; do \
	test -e $x/.keep && continue ; \
	install -m 0755 -d $x || exit $? ; \
	touch $x/.keep || echo "ignoring touch failure; mounted fs?" ; \
done
/usr/bin/install: cannot change permissions of ‘/sys’: Read-only file system
make: *** [Makefile:67: layout-dirs] Error 1
```